samza-core/src/main/java/org/apache/samza/zk/ZkBarrierForVersionUpgrade.java

```diff
@@ -40,74 +55,71 @@
+  public ZkBarrierForVersionUpgrade(ZkUtils zkUtils, ScheduleAfterDebounceTime debounceTimer, String version, List<String> processorsNames) {
     this.zkUtils = zkUtils;
     keyBuilder = zkUtils.getKeyBuilder();
 
     barrierPrefix = keyBuilder.getJobModelVersionBarrierPrefix();
     this.debounceTimer = debounceTimer;
+
+    barrierPath = String.format("%s/barrier_%s", barrierPrefix, version);
+    barrierDonePath = String.format("%s/barrier_done", barrierPath);
+    barrierProcessors = String.format("%s/barrier_processors", barrierPath);
+
+    this.version = version;
+    this.processorsNames = processorsNames;
+
+    zkUtils.makeSurePersistentPathsExists(new String[]{barrierPrefix, barrierPath, barrierProcessors, barrierDonePath});
   }
 
   /**
    * set the barrier for the timer. If the timer is not achieved by the timeout - it will fail
    * @param version for which the barrier is created
    * @param timeout - time in ms to wait
    */
-  public void setTimer(String version, long timeout) {
-    debounceTimer.scheduleAfterDebounceTime("VersionUpgradeTimeout", timeout, ()->timerOff(version));
+  public void setTimer(final String version, final long timeout, final Stat currentStatOfBarrierDone) {
+    debounceTimer.scheduleAfterDebounceTime("VersionUpgradeTimeout", timeout, ()->timerOff(version, currentStatOfBarrierDone));
   }
 
   protected long getBarrierTimeOutMs() {
     return BARRIER_TIMED_OUT_MS;
   }
 
-  private void timerOff(String version) {
-    // check if barrier has finished
-    final String barrierPath = String.format("%s/barrier_%s", barrierPrefix, version);
-    final String barrierDonePath = String.format("%s/barrier_done", barrierPath);
-    Stat stat = new Stat();
-    String done = zkUtils.getZkClient().<String>readData(barrierDonePath, stat);
-    if (done != null && done.equals(BARRIER_DONE))
-      return; //nothing to do
-
-    while (true) {
-      try {
-        // write a new value if no one else did, if the value was changed since previous reading - retry
-        zkUtils.getZkClient().writeData(barrierDonePath, "TIMED_OUT", stat.getVersion());
-        return;
-      } catch (Exception e) {
-        // failed to write, try read/write again
-        LOG.info("Barrier timeout write failed");
-        done = zkUtils.getZkClient().<String>readData(barrierDonePath, stat);
-        if (done.equals(BARRIER_DONE))
-          return; //nothing to do
+  private void timerOff(final String version, final Stat currentStatOfBarrierDone) {
+    try {
+      // write a new value "TIMED_OUT", if the value was changed since previous value, make sure it was changed to "DONE"
+      zkUtils.getZkClient().writeData(barrierDonePath, BARRIER_TIMED_OUT, currentStatOfBarrierDone.getVersion());
+    } catch (Exception e) {
+      // failed to write, make sure the value is "DONE"
+      LOG.info("Barrier timeout write failed");
+      String done = zkUtils.getZkClient().<String>readData(barrierDonePath);
```

Comment:
The only valid case for the write to fail is that the barrier was reached, and the value is DONE. In this case we do nothing. If we successfully write "TIMED_OUT", this will cause all the participants to abandon the barrier.